### ISSUE TITLE
Feature request: Inlay hint

### ISSUE THREAD

**Issue by fecet** [python-lsp/python-lsp-server#397]
python still don't have avaliable language server support inlay hint, hope pylsp can be the hero :smile:

**Comment by ccordoba12** [comment]
Hey @fecet, thanks for reporting. What do you mean by this?

**Comment by fecet** [comment]
[The inlay hint feature](https://github.com/microsoft/language-server-protocol/issues/956) is stabilized in latest [LSP Spec 3.17](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/), and implemented in VSCode and Neovim.

Currently pylsp doesn't support that

**Comment by Amar1729** [comment]
Just as a note - inlay hints are supported by [coc-pyright](https://github.com/fannheyward/coc-pyright/blob/0d88f8ff7c7afcf812497a5f283b43b74df35737/src/features/inlayHints.ts) and VSCode's [pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance), but notably [not pyright](https://github.com/microsoft/pyright/issues/4325) (although I don't use either pylance, which requires VSCode, or coc-pyright).

**Comment by markis** [comment]
[based-pyright](https://github.com/DetachHead/basedpyright) supports [inlay hints](https://github.com/DetachHead/basedpyright?tab=readme-ov-file#inlay-hints)

**Comment by baco** [comment]
> Just as a note - inlay hints are supported by [coc-pyright](https://github.com/fannheyward/coc-pyright/blob/0d88f8ff7c7afcf812497a5f283b43b74df35737/src/features/inlayHints.ts) and VSCode's [pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance), but notably [not pyright](https://github.com/microsoft/pyright/issues/4325).

I don't know `coc-pyright` approach, but:

* VSCode's Pylance is not possible to make it work.  It has proprietary licensing and does some checks against Microsoft's marketplace, that even VSCodium staff are having issues in giving the support for running outside VSCode.
* VSCode's Pyright won't have the feature.  Microsoft has chopped that functionality out, in favor of adding that to Pylance.  Pylance and Pyright are, some how complimentary.  Pyright does only type validation.

As @markis said, BasedPyright project supports that. And it's even more accessible for Python developers, since doesn't force users to install NodeJS (like coc-pyright o pyright require).  It does install NodeJS behind the curtains though, but it's all encapsulated in a virtual environment, transparent to the Python developer.

**Comment by Amar1729** [comment]
Yeah, at this point i would definitely recommend basedpyright. It didn't exist at the time I made my comment lol

**Comment by claytharrison** [comment]
So there's a plugin that provides this (https://github.com/BoringStuffInc/pylsp-inlay-hints), but as an *experimental* capability. At least in my case, eglot for Emacs looks for `inlayHintProvider` as a regular capability, so it doesn't work.

I would like to adapt the plugin to do that, but I can't tell if there's any way for plugins to provide a new *proper* capability to pylsp. I only see a hookspec for experimental ones. I'll keep digging, but does anyone have any hints on this?

EDIT:

All I have to do to get `pylsp-inlay-hints` working with eglot is to add `inlayHintProvider` to pylsp's capabilities in `python_lsp.py`, i.e.
```python
    def capabilities(self):
        server_capabilities = {
            ...
            "inlayHintProvider": {
                "resolveProvider": False,
                "workDoneProgress": True,
            },
            ...
        }
```
